src/ImageSharp/ColorSpaces/Conversion/Implementation/Icc/Calculators/ClutCalculator.cs

```diff
@@ -0,0 +1,167 @@
+                    result = this.Interpolate(value.X);
+                    break;
+                case 2:
+                    result = this.Interpolate(value.X, value.Y);
+                    break;
+                case 3:
+                    result = this.Interpolate(value.X, value.Y, value.Z);
+                    break;
+                case 4:
+                    result = this.Interpolate(value.X, value.Y, value.Z, value.W);
+                    break;
+
+                default:
+                    throw new InvalidOperationException();
+            }
+
+            switch (result.Length)
+            {
+                case 1:
+                    return new Vector4(result[0], 0, 0, 0);
+                case 2:
+                    return new Vector4(result[0], result[1], 0, 0);
+                case 3:
+                    return new Vector4(result[0], result[1], result[2], 0);
+                case 4:
+                    return new Vector4(result[0], result[1], result[2], result[3]);
+
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private int[] CalculateIndexFactor(int inputCount, byte[] gridPointCount)
+        {
+            this.indexFactor = new int[inputCount];
+            int gpc = 1;
+            for (int j = inputCount - 1; j >= 0; j--)
+            {
+                this.indexFactor[j] = gpc * (gridPointCount[j] - 1);
+                gpc *= gridPointCount[j];
+            }
+
+            return this.indexFactor;
+        }
+
+        private float[] Interpolate(float a)
+        {
+            return this.Interpolate(new float[] { a });
+        }
+
+        private float[] Interpolate(float a, float b)
+        {
+            return this.Interpolate(new float[] { a, b });
+        }
+
+        private float[] Interpolate(float a, float b, float c)
+        {
+            return this.Interpolate(new float[] { a, b, c });
+        }
+
+        private float[] Interpolate(float a, float b, float c, float d)
+        {
+            return this.Interpolate(new float[] { a, b, c, d });
+        }
+
+        private float[] Interpolate(float[] values)
+        {
+            float[][] nodes = new float[this.nodeCount][];
+            for (int i = 0; i < nodes.Length; i++)
+            {
+                int index = 0;
+                for (int j = 0; j < values.Length; j++)
+                {
+                    float fraction = 1f / (this.gridPointCount[j] - 1);
+                    int position = (int)(values[j] / fraction);
+                    if (((i >> j) & 1) == 1)
+                    {
+                        position += 1;
+                    }
+
+                    index += (int)((this.indexFactor[j] * (position * fraction)) + 0.5f);
+                }
+
+                nodes[i] = this.lut[index];
+            }
+
+            float[] factors = new float[this.nodeCount];
+            for (int i = 0; i < factors.Length; i++)
```

Comment:
This looks like it could be vectorized but I could be wrong, @antonfirsov  What do you think?

Reply:
Depends on what `if (((i >> j) & 1) == 1)` does. We need to eliminate branches inside loops for vectorization.